### PR TITLE
Friends can now be managed

### PR DESCRIPTION
**Pull request by eprei**
Now you can:

- delete friends
- accept friend requests sent by another user
- delete requests that you have sent yourself
- add new friends by choosing them from a list of all users who are not our friends

To test it:

- make clean-database
- make
- login with 42
- http://localhost:8080/api/ (create dummy data)
- go to profile page
- accept, delete, add friends

You should have a pending friend request from another user along with other things

### CODE REVIEW

File: frontend/src/components/profile/OtherUsers.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    ></div>
+                    <div className={styles.nameAndStatus}>
+                        <CustomLink to={`/user/${nickname}`}>
+                            <h3>{nickname}</h3>
+                        </CustomLink>
+                    </div>
+                </>
+            ) : (
+                <div></div>
+            )}
```
Comment (RobinBurri) [inline frontend/src/components/profile/OtherUsers.tsx:65]:
```suggestion
                   {!successfullyDone && (
                <>
                    <div>
                        <ClickableIcon
                            icon={IconAddFriend}
                            onClick={() => sendFriendRequest(id)}
                        ></ClickableIcon>
                    </div>
                    <div
                        className={styles.profilePicture}
                        style={profilePictureStyle}
                    ></div>
                    <div className={styles.nameAndStatus}>
                        <CustomLink to={`/user/${nickname}`}>
                            <h3>{nickname}</h3>
                        </CustomLink>
                    </div>
                </>
            )}
```

Comment (RobinBurri) [reply frontend/src/components/profile/OtherUsers.tsx:65]: 1) Simplify, no need to return an empty div if successfullyDone.
2) span with title is useless (I don't see it anywhere on the page)

File: frontend/src/components/profile/Friend.tsx [hunk 2]
```diff
@@ -74,36 +59,113 @@ const Friend = ({ name, picture, status, isFriend }: FriendProps) => {
...
+                            <h3>{nickname}</h3>
+                        </CustomLink>
+                        <p className={`${styles.status} ${statusColorClass}`}>
+                            {status}
+                        </p>
+                    </div>
+                </>
+            ) : (
+                <div></div>
+            )}
```
Comment (RobinBurri) [inline frontend/src/components/profile/Friend.tsx:168]:
```suggestion
              {!successfullyDone && (
                <>
                    <div>
                        {isPending && !createdByMe && (
                            <ClickableIcon
                                icon={IconRemoveFriend}
                                onClick={() => removeFriendship(id)}
                            ></ClickableIcon>
                        )}

                        <ClickableIcon
                            icon={
                                isPending
                                    ? createdByMe
                                        ? IconRemoveFriend
                                        : IconAcceptFriend
                                    : IconRemoveFriend
                            }
                            onClick={
                                isPending
                                    ? createdByMe
                                        ? () => removeFriendship(id)
                                        : () => acceptFriendship(id)
                                    : () => removeFriendship(id)
                            }
                        ></ClickableIcon>
                    </div>
                    <div
                        className={styles.profilePicture}
                        style={profilePictureStyle}
                    ></div>
                    <div className={styles.nameAndStatus}>
                        <CustomLink to={`/user/${nickname}`}>
                            <h3>{nickname}</h3>
                        </CustomLink>
                        <p className={`${styles.status} ${statusColorClass}`}>
                            {status}
                        </p>
                    </div>
                </>
            )}
```

Comment (RobinBurri) [reply frontend/src/components/profile/Friend.tsx:168]: 1) Simplify, no need to return an empty div if successfullyDone.
2) span with title is useless (I don't see it anywhere on the page)

File: frontend/src/components/profile/OtherUsers.tsx (hunk 1 above)
Comment (eprei) [reply frontend/src/components/profile/OtherUsers.tsx:65]: Good. Thanks for the improvements you have suggested. I find them very pertinent and educational.

File: frontend/src/components/profile/Friend.tsx (hunk 2 above)
Comment (eprei) [reply frontend/src/components/profile/Friend.tsx:168]: Good. Thanks for the improvements you have suggested. I find them very pertinent and educational.

**Review by RobinBurri — APPROVED** [review]
Nice work

**Comment by eprei** [comment]
@RobinBurri Thanks to you for your improvements and advice

**Review by shaolin-peanut — APPROVED** [review]
Big work, congrats!

**Review by t-h2o — COMMENTED** [review]
:broom:

File: backend/src/app.service.ts [hunk 3]
```diff
@@ -118,21 +180,70 @@ export class AppService {
...
 
         // Create friendships
-        const users = [user1, user2, user3, user4, user5, user6]
+        const users = [userMe, user1, user2, user3, user4, user5, user6]
 
-        for (const user of users) {
-            const friends = users.filter((p) => p.id !== user.id).slice(0, 3)
+        // for (const user of users) {
+        //     const friends = users.filter((p) => p.id !== user.id).slice(0, 3)
+        //     let j: number = 0
```
[the remark this answers is not in the document]
Comment (eprei) [reply backend/src/app.service.ts:187]: To develop this feature, I have manually created relationships one by one, that's why I commented this part of the code. However, i did not want to delete these lines as it is the method of automatic creation of friendships previously created by Robin. As I don't know if the rest of the team is basing their development on these friendships I didn't want to delete it.

**Review by t-h2o — APPROVED** [review]
LGTM

**Pull request merged**